### DISCUSSION TITLE
Issue with MOOSE Nonlinear System Assembly

### DISCUSSION THREAD

**Discussion by hityyds** [idaholab/moose#26398] [Q&A General]
I'm curious about how MOOSE assembles the nonlinear systems it needs to solve. If I provide two completely unrelated problems in the same inputdeck.i, will MOOSE combine the nonlinear equations of these two problems into one large system of equations and then call Petsc to solve it, or will it independently call Petsc to solve each problem? 

In the case of two coupled problems, if I use auxvariables to decouple them (updating auxvariables with the values from the previous iteration after each linear iteration and then using auxvariables to replace the coupled variables in the next iteration), will MOOSE still combine the two problems into one large nonlinear system? 

In fact, I have already attempted to use auxvariables to completely decouple the equations between different variables. I thought this would significantly reduce the memory usage of MOOSE because, after decoupling, the problem to be solved would shift from one extensive system of equations to multiple smaller systems. I believed this would greatly reduce the dimension of the coefficient matrix that needs to be stored. However, the test results revealed that after decoupling, MOOSE's memory usage remained almost unchanged, but the computation time decreased by nearly half. 

If different problems are combined into one set, why does MOOSE have multiple nl_sys? My understanding is that one nonlinear system corresponds to one set of nonlinear equations.
```
inline void
FEProblemBase::setCurrentNonlinearSystem(const unsigned int nl_sys_num)
{
  mooseAssert(nl_sys_num < _nl.size(),
              "System number greater than the number of nonlinear systems");
  _current_nl_sys = _nl[nl_sys_num].get();
}
```

**Comment by GiudGiud** [comment]
Hello

The default is combining and solving both equations together.

If you assign the variables to different systems, you can now solve system separately. This is a current, ongoing, development. We have a separate solve between two systems in an Executioner test object that's it.

Your understanding is correct. Different systems are solved separately. 
However what you did with auxiliary variables provided some degree of separation between the solves, but not to the extent of "different matrices" or "solved separately". There was still some coupling every time the auxkernel was run and it updated the variable.

Guillaume

**Reply by hityyds** [reply]
I understand that there is still some coupling between different problems even after using auxvariables. However, my curiosity lies in whether it is feasible to utilize auxvariables to allocate different variables to distinct systems. The idea is to update auxvariables by exchanging information between different systems only after separately completing one iteration for all systems (not multiapp). Additionally, how can one go about assigning different variables to different systems?

**Reply by lindsayad** [reply]
You can look at https://github.com/idaholab/moose/tree/next/test/tests/misc/multiple-nl-systems
